akka_microservices/ibex35Microservice/src/main/scala/com/tapiporla/microservices/retrievers/common/stats/StatsGenerator.scala

```diff
@@ -13,79 +12,70 @@ import scala.annotation.tailrec
-    * @param data
-    * @param previousData
-    * @return
-    */
-  def generateStatsFor(data: Seq[Data], previousData: Seq[Data]): Seq[Stat] = {
-    val dataByDate = data.sortBy(_._1.toDate)
-    val previousByDate = data.sortBy(_._1.toDate)
-
-    generateMM200From(dataByDate, previousByDate takeRight MM200._2 - 1) ++
-      generateMM100From(dataByDate, previousByDate takeRight MM100._2 -1) ++
-      generateMM40From(dataByDate, previousByDate takeRight MM40._2 - 1) ++
-      generateMM20From(dataByDate, previousByDate takeRight MM20._2 -1)
+object StatsGenerator extends LazyLogging {
+
+  object MMDefinition {
+
+    def from(numberOfItems: Int): MMDefinition =
+      MMDefinition(s"MM$numberOfItems", numberOfItems)
+
   }
 
-  def generateMM200From(data: Seq[Data], previousData: Seq[Data]): Seq[Stat] =
-    generateMM(data, previousData, MM200._2) map {item => (item._1, MM200._1, item._2)}
+  case class MMDefinition(identifier: String, numberOfItems: Int)
 
-  def generateMM100From(data: Seq[Data], previousData: Seq[Data]): Seq[Stat] =
-    generateMM(data, previousData, MM100._2) map {item => (item._1, MM100._1, item._2)}
+  case class StatGenerated(statType: String, date: DateTime, value: BigDecimal)
+  case class StatDataInput(date: DateTime, value: BigDecimal)
 
-  def generateMM40From(data: Seq[Data], previousData: Seq[Data]): Seq[Stat] =
-    generateMM(data, previousData, MM40._2) map {item => (item._1, MM40._1, item._2)}
 
-  def generateMM20From(data: Seq[Data], previousData: Seq[Data]): Seq[Stat] =
-    generateMM(data, previousData, MM20._2) map {item => (item._1, MM20._1, item._2)}
+  def generateMultipleMMs(
+                           initial: Seq[StatDataInput],
+                           previousData: Seq[StatDataInput],
+                           mmsToProcess: Seq[MMDefinition]
+                         ) = {
 
-  def generateMM(initial: Seq[Data], previousData: Seq[Data], number: Int): Seq[Data] = {
+
+    val dataByDate = initial.sortBy(_.date.toDate)
+    val previousByDate = previousData.sortBy(_.date.toDate)
+
+    mmsToProcess flatMap { mm =>
+      generateMM(dataByDate, previousByDate takeRight mm.numberOfItems - 1, mm)
+    }
+  }
+
+
+  protected def generateMM(
+                  initial: Seq[StatDataInput],
+                  previousData: Seq[StatDataInput],
+                  mm: MMDefinition
+                ): Seq[StatGenerated] = {
 
     @tailrec
-    def helper(pending: Seq[Data], passed: Seq[Data], accum: Seq[Data]): Seq[Data] = {
-      if(pending.isEmpty)
+    def helper(pending: Seq[StatDataInput], passed: Seq[StatDataInput], accum: Seq[StatGenerated]): Seq[StatGenerated] = {
+      if (pending.isEmpty)
         accum
       else {
         val newChunk =
-          if(passed.length < number) //(number == initial length elements)
-             passed :+ pending.head
+          if (passed.length < mm.numberOfItems) //(number == initial length elements)
+            passed :+ pending.head
```

Comment:
Mmm just thinking. Prepending has O(1) complexity while appending has O(n*log(n)).
I think it's much better to prepend and in base case, return ```accum.reverse```.
That way we reduce complexity to O(n)

Reply:
I will convert that to a Queue. That is really the structure that is more similar (we are just appending in a limit sized Queue)

Reply:
I've made metrics, and the difference between one implementation and other in our case, is not significative... I will maintain the Seq, and if we need to optimize that, we can change the algorithm (for example to calculate in parallel the different MM instead of sequencially)